closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1070,23 +924,604 @@ boolean isRestrictedProviderAllowed(String providerName) {
+            }
+            setProperty("descNumber", profileID + ".desc.number", allInfo);
+            setProperty("descPolicy", profileID + ".desc.policy", allInfo);
+            setProperty("descSunsetDate", profileID + ".desc.sunsetDate", allInfo);
+
+            setProperty("jdkTlsDisabledNamedCurves",
+                    profileID + ".tls.disabledNamedCurves", allInfo);
+            setProperty("jdkTlsDisabledAlgorithms",
+                    profileID + ".tls.disabledAlgorithms", allInfo);
+            setProperty("jdkTlsEphemeralDHKeySize",
+                    profileID + ".tls.ephemeralDHKeySize", allInfo);
+            setProperty("jdkTlsLegacyAlgorithms",
+                    profileID + ".tls.legacyAlgorithms", allInfo);
+            setProperty("jdkCertpathDisabledAlgorithms",
+                    profileID + ".jce.certpath.disabledAlgorithms", allInfo);
+            setProperty("jdkSecurityLegacyAlgorithms",
+                    profileID + ".jce.legacyAlgorithms", allInfo);
+            setProperty("keyStoreType",
+                    profileID + ".keystore.type", allInfo);
+            setProperty("keyStore",
+                    profileID + ".javax.net.ssl.keyStore", allInfo);
+
+            setProperty("jdkSecureRandomProvider",
+                    profileID + ".securerandom.provider", allInfo);
+            setProperty("jdkSecureRandomAlgorithm",
+                    profileID + ".securerandom.algorithm", allInfo);
+            setProperty("jdkFipsMode",
+                    profileID + ".fips.mode", allInfo);
+
+            if (debug != null) {
+                debug.println("\tProperties of restricted security profile successfully loaded.");
+            }
+        }
+
+        private void setConstraints(String providerName, String providerInfo, boolean providerChanged) {
+            if (debug != null) {
+                debug.println("\t\tLoading constraints for security provider: " + providerName);
+            }
+
+            List<Constraint> constraints = new ArrayList<>();
+
+            providerInfo = providerInfo.replaceAll("\\s+", "");
+
+            // Check whether constraints are specified for this provider.
+            Pattern p = Pattern.compile("\\[.+\\]");
+            Matcher m = p.matcher(providerInfo);
+            if (!m.find()) {
+                if (debug != null) {
+                    debug.println("\t\t\tNo constraints for security provider: " + providerName);
+                }
+                providerConstraints.put(providerName, constraints);
+                return;
+            }
+
+            // Check whether constraints are properly specified.
+            p = Pattern.compile(
+                "\\["
+                + "([\\+\\-]?)"                     // option to append or remove
+                + "(\\{"                            // at least one constraint
+                    + "\\w+,"
+                    + "[A-Za-z0-9\\.\\-/_]+,"
+                    + "[A-Za-z0-9\\=\\*\\|\\.:]+"
```

Comment:
Please avoid escapes where they're not needed (within `[]`, `.`, `=`, `*`, `|`, `:` are treated literally, and `-` (if the first or last character) need not be escaped). In fact, the javadoc for `Pattern` says:
> It is an error to use a backslash prior to any alphabetic character that does not denote an escaped construct

When I originally suggested decomposing patterns, I was thinking the parts could be named, like this:
```java
            final String typeRE = "\\w+";
            final String algoRE = "[A-Za-z0-9./_-]+";
            final String attrRE = "[A-Za-z0-9=*|.:]+";
            final String consRE = "\\{(" + typeRE + "),(" + algoRE + "),(" + attrRE + ")\\}";

            p = Pattern.compile(
                "\\["
                + "([+-]?)"             // option to append or remove
                + consRE                // at least one constraint
                + "(," + consRE + ")*"  // more constraints [optional]
                + "\\]");
```
With the parentheses included in `consRE` above, that pattern can directly be reused to parse constraints:
```java
            // Parse all provided constraints.
            p = Pattern.compile(consRE);
```

Reply:
Makes sense. I didn't know we don't need to escape special characters when in `[]`. Changed to suggestion.

Reply:
Most characters between `[` and `]` don't need to be escaped, but some do - see https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html.